### PR TITLE
Change auditor interface to avoid holding key and values in the producer before the message is acked.

### PR DESCRIPTION
**Pull request by becketqin**

### CODE REVIEW

File: src/main/java/com/linkedin/kafka/clients/auditing/Auditor.java [hunk 1]
```diff
@@ -36,22 +36,26 @@
...
-   * for each record it sends, no matter if the record is sent successfully or failed.
+   * Get the custom audit information from the key and value of the record.
+   * This method helps the producer avoid holding the key and value until the message sending is completed.
    *
-   * This method may be called from multiple threads, so the implementation must be thread safe.
+   * @param key the key of the record.
+   * @param value the value of the record.
+   * @return the custom audit information.
+   */
+  Object getCustomAuditInfo(K key, V value);
```
Comment (radai-rosenblatt) [inline src/main/java/com/linkedin/kafka/clients/auditing/Auditor.java]: i think Audit Token is a better term

File: src/main/java/com/linkedin/kafka/clients/auditing/Auditor.java [hunk 2]
```diff
@@ -36,22 +36,26 @@
...
+   * @param topic The topic of the record.
+   * @param timestamp The timestamp of the record.
    * @param messageCount The number of messages to record.
    * @param bytesCount The number of bytes to record.
-   * @param auditType The type of the event to audit.
    */
-  void record(String topic,
-              K key,
-              V value,
+  void record(Object customAuditInfo,
```
Comment (radai-rosenblatt) [inline src/main/java/com/linkedin/kafka/clients/auditing/Auditor.java]: auditToken ? :-)

File: src/main/java/com/linkedin/kafka/clients/auditing/abstractimpl/CountingAuditStats.java [hunk 3]
```diff
@@ -47,10 +47,10 @@ public void update(Object auditKey, long messageCount, long bytesCount) {
         throw new IllegalStateException("Stats has been closed. The caller should get the new AuditStats and retry.");
       }
 
-      AuditInfo statsForTopic = _stats.get(auditKey);
+      AuditingCounts statsForTopic = _stats.get(auditKey);
       if (statsForTopic == null) {
-        statsForTopic = new AuditInfo();
-        AuditInfo prev = _stats.putIfAbsent(auditKey, statsForTopic);
+        statsForTopic = new AuditingCounts();
+        AuditingCounts prev = _stats.putIfAbsent(auditKey, statsForTopic);
```
Comment (radai-rosenblatt) [inline src/main/java/com/linkedin/kafka/clients/auditing/abstractimpl/CountingAuditStats.java]: computeIfAbsent would be more efficient (wont allocate new object if exists)

File: src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java [hunk 4]
```diff
@@ -194,6 +194,7 @@ private LiKafkaProducerImpl(LiKafkaProducerConfig configs,
       String topic = producerRecord.topic();
       K key = producerRecord.key();
       V value = producerRecord.value();
+      Object auditInfo = _auditor.getCustomAuditInfo(key, value);
```
Comment (radai-rosenblatt) [inline src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java]: token :-D

File: src/main/java/com/linkedin/kafka/clients/auditing/abstractimpl/AbstractAuditor.java [hunk 5]
```diff
@@ -144,6 +145,8 @@ public void run() {
             LOG.error("Auditor encounter exception.", e);
           }
         }
+      } catch (Throwable t) {
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kafka/clients/auditing/abstractimpl/AbstractAuditor.java]: Don't catch Throwable catch Exception.  There's almost no conditions which are recoverable when a Throwable is thrown that is not an Exception.

File: src/main/java/com/linkedin/kafka/clients/auditing/abstractimpl/AbstractAuditor.java [hunk 6]
```diff
@@ -44,13 +44,14 @@
...
- *   {@link #getAuditKey(String, Object, Object, Long, Long, Long, com.linkedin.kafka.clients.auditing.AuditType)}
+ *   {@link #getAuditKey(Object, String, Long, Long, Long, AuditType)}
+ *   {@link #getCustomAuditInfo(Object, Object)}
  * </pre>
  *
  * <p>
  * For users who wants to have customized configurations, they may override the method:
  * <pre>
- *   public void configure(Map&lt;String, ?&gt; configs);
+ *   public void configure(Map<String, ?> configs);
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kafka/clients/auditing/abstractimpl/AbstractAuditor.java]: This breaks the HTML escaping.

File: src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java [hunk 7]
```diff
@@ -298,43 +299,40 @@ private void prepClose() {
...
       _userCallback = userCallback;
     }
 
     @Override
     public void onCompletion(RecordMetadata recordMetadata, Exception e) {
       if (e != null) {
-        LOG.error(String.format("Unable to send event %s with key %s and message id %s to kafka topic %s",
-            _value.toString(), (_key != null) ? _key : "[none]",
+        LOG.error(String.format("Unable to send event %s with message id %s to kafka topic %s",
+                                _auditInfo == null ? "[No Custom Info]" : _auditInfo,
```
Comment (smccauliff) [inline src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java]: No audit token?

Comment (radai-rosenblatt) [inline src/main/java/com/linkedin/kafka/clients/producer/LiKafkaProducerImpl.java]: got repetitive :-)

**Pull request merged**